src/libcore/slice.rs

```diff
@@ -643,6 +649,18 @@ macro_rules! slice_offset {
     }};
 }
 
+// Use this to do `$ptr - $by`, where `$by` is non-negative.
+macro_rules! slice_sub_offset {
+    ($ptr:expr, $by:expr) => {{
+        let ptr = $ptr;
+        if size_from_ptr(ptr) == 0 {
+            transmute(ptr as usize - $by)
```

Comment:
I have kept the changes in the form presented here because I still believe this code is easier to read than that change that you suggest. (It is an entirely subjective opinion.)

I freely admit that I have not done the full review of the `core::slice` code, and thus I have no argument against the point you make about hypothetical fixes to #25016.  If it comes to that, then we can revert back to a macro of the form you suggest when #25016 gets fixed.

For now, however, my main concern is getting `make check` working again on debug builds, and this was the most obviously correct way for me to go about doing it.